test/PushNotifications/PushNotificationsLongRunningTask.Msix/PushNotificationsLongRunningTask.Msix.vcxproj

```diff
@@ -63,68 +63,37 @@
-    <NMakePreprocessorDefinitions>WIN32;NDEBUG;$(NMakePreprocessorDefinitions)</NMakePreprocessorDefinitions>
-    <OutDir>$(OutDir)</OutDir>
-    <IntDir>$(IntDir)</IntDir>
-    <NMakeIncludeSearchPath>$(SolutionDir)tools</NMakeIncludeSearchPath>
-  </PropertyGroup>
-  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|ARM64'">
-    <NMakeBuildCommandLine>nmake /f makefile.mak ProjectDir=$(ProjectDir) SolutionDir=$(SolutionDir) OutDir=$(OutDir) TargetName=$(TargetName) WindowsAppSDKBuildPipeline=$(WindowsAppSDKBuildPipeline) </NMakeBuildCommandLine>
-    <NMakeOutput>$(OutputPath)PushNotificationsLongRunningTask.msix</NMakeOutput>
-    <NMakeCleanCommandLine>nmake /f makefile.mak ProjectDir=$(ProjectDir) SolutionDir=$(SolutionDir) OutDir=$(OutDir) TargetName=$(TargetName) WindowsAppSDKBuildPipeline=$(WindowsAppSDKBuildPipeline) clean</NMakeCleanCommandLine>
-    <NMakeReBuildCommandLine>nmake /f makefile.mak ProjectDir=$(ProjectDir) SolutionDir=$(SolutionDir) OutDir=$(OutDir) TargetName=$(TargetName) WindowsAppSDKBuildPipeline=$(WindowsAppSDKBuildPipeline) rebuild</NMakeReBuildCommandLine>
-    <NMakePreprocessorDefinitions>WIN32;NDEBUG;$(NMakePreprocessorDefinitions)</NMakePreprocessorDefinitions>
-    <OutDir>$(OutDir)</OutDir>
-    <IntDir>$(IntDir)</IntDir>
-    <NMakeIncludeSearchPath>$(SolutionDir)tools</NMakeIncludeSearchPath>
-  </PropertyGroup>
-  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
-    <NMakeBuildCommandLine>nmake /C /f makefile.mak ProjectDir=$(ProjectDir) SolutionDir=$(SolutionDir) OutDir=$(OutDir) TargetName=$(TargetName) WindowsAppSDKBuildPipeline=$(WindowsAppSDKBuildPipeline) </NMakeBuildCommandLine>
-    <NMakeOutput>$(OutputPath)PushNotificationsLongRunningTask.msix</NMakeOutput>
-    <NMakeCleanCommandLine>nmake /C /f makefile.mak ProjectDir=$(ProjectDir) SolutionDir=$(SolutionDir) OutDir=$(OutDir) TargetName=$(TargetName) WindowsAppSDKBuildPipeline=$(WindowsAppSDKBuildPipeline) clean</NMakeCleanCommandLine>
-    <NMakeReBuildCommandLine>nmake /C /f makefile.mak ProjectDir=$(ProjectDir) SolutionDir=$(SolutionDir) OutDir=$(OutDir) TargetName=$(TargetName) WindowsAppSDKBuildPipeline=$(WindowsAppSDKBuildPipeline) rebuild</NMakeReBuildCommandLine>
-    <NMakePreprocessorDefinitions>NDEBUG;$(NMakePreprocessorDefinitions)</NMakePreprocessorDefinitions>
-    <OutDir>$(OutDir)</OutDir>
-    <IntDir>$(IntDir)</IntDir>
-    <NMakeIncludeSearchPath>$(SolutionDir)tools</NMakeIncludeSearchPath>
-  </PropertyGroup>
-  <ItemDefinitionGroup>
-  </ItemDefinitionGroup>
+  <ItemGroup>
+    <MakeMsixInputAssets Include="$(MSBuildThisFileDirectory)\Assets\*" />
+    <MakeMsixInputs Include="$(MSBuildThisFileDirectory)appxmanifest.xml" />
+    <MakeMsixInputs Include="$(OutDir)..\WindowsAppRuntime_DLL\Microsoft.Internal.FrameworkUdk.dll" />
```

Comment:
Avoid this pattern please every chance you get.

"$(OutDir)..\WindowsAppRuntime_DLL" == "$(BaseOutputPath)\WindowsAppRuntime_DLL"

Relative paths are terrible on build systems.  Not only does it make refactoring hard, but causes hidden bugs in the code base due to standard file names for headers.  When you move files in a refactor you can accidentally build clean, when you are including the wrong files.  It leads to very hard to diagnose problems.  If you can't find a variable that allows avoiding relative directory usage, let me know and I'll gladly help you find the right solution.

Reply:
What's the recommended macros? The important ones I know:

* RepoRoot = \
* OutDir = $(RepoRoot)\BuildOutput\$(Configuration)\$(PlatformTarget)\$(MSBuildProjectName)
* IntDir = $(RepoRoot)\$(Configuration)\$(PlatformTarget)\$(MSBuildProjectName)

At least, that's what I thought they are in usual VS projects. Am I mistaken?

Looking at EnsureOutputLayout.props...

* BaseOutputPath = $(RepoRoot)\BuildOutput\$(Configuration)\$(PlatformTarget)\
* OutDir = $(BaseOutputPath)
* OutputPath = $(OutDir)\$(MSBuildProjectName)
* TargetDir == $(OutDir)

Why do we have TargetDir? Is it a VS macro or one of our own invention?

OutDir is the common dir and OutputPath is a project's output dir?

* BaseIntermediateOutputPath = $(RepoRoot)\obj\$(Configuration)\$(PlatformTarget)\$(MSBuildProjectName)
* IntermediateOutputPath = $(BaseIntermediateOutputPath)

Why no IntDir?

Naming is confusing -- OutputBlah vs IntermediateOutputBlah. Took me a while to figure out the latter is a peer of the former and not a child / dependent. Would be better as OutputBlah and IntermediateBlah. Any reason we can't do that?



Also, what's the recommended syntax for one project o reference another project's OutDir? You suggest `$(OutDir)..\foo` is bad. Do what instead? `$(RepoRoot)\BuildOutput\$(Configuration)\$(PlatformTarget)\foo`? Is there a macro for `$(RepoRoot)\BuildOutput\$(Configuration)\$(PlatformTarget)\`?

According to EnsureOutputLayout.props that _is_ OutDir, but per actually building projects OutDir is what EnsureOutputLayout.props calls OutputPath. So I'm puzzled how it winds up in that state, and of course the $64,000 question - what should we change?